kafka-rest/src/main/java/io/confluent/kafkarest/backends/kafka/KafkaModule.java

```diff
@@ -37,7 +39,33 @@ public KafkaModule(KafkaRestContext context) {
 
   @Override
   protected void configure() {
-    // Reuse the AdminClient being constructed in KafkaRestContext.
-    bind(context.getAdmin()).to(Admin.class);
+    // Reuse the AdminClient being constructed in KafkaRestContext. The request-scope is needed
+    // because the Admin creation logic uses request-scoped auth information. KafkaRestContext
+    // itself is a proxied instance, so the result will be cached there appropriately, after it is
+    // created. See KafkaRestContextProvider and ContextInvocationHandler for more information.
+    bindFactory(new AdminFactory(context))
+        .to(Admin.class)
+        .in(RequestScoped.class)
+        .proxy(true)
+        .proxyForSameScope(true);
+  }
+
+  private static final class AdminFactory implements Factory<Admin> {
+
+    private final KafkaRestContext context;
+
+    private AdminFactory(KafkaRestContext context) {
+      this.context = Objects.requireNonNull(context);
+    }
+
+    @Override
+    public Admin provide() {
+      return context.getAdmin();
+    }
+
+    @Override
+    public void dispose(Admin instance) {
+      // Do nothing.
```

Comment:
Where is the AdminClient closed?

Reply:
AdminFactory has no ownership of Admin. It only proxies it to KafkaRestContext. KafkaRestContext owns and disposes of it.

https://github.com/confluentinc/kafka-rest/blob/master/kafka-rest/src/main/java/io/confluent/kafkarest/DefaultKafkaRestContext.java#L112
https://github.com/confluentinc/kafka-rest/blob/master/kafka-rest/src/main/java/io/confluent/kafkarest/AdminClientWrapper.java#L157